modules/monitoring-resolving-the-kubepersistentvolumefillingup-alarm-firing-for-prometheus.adoc

```diff
@@ -0,0 +1,67 @@
+// Module included in the following assemblies:
+//
+// * monitoring/troubleshooting-monitoring-issues.adoc
+// * support/troubleshooting/investigating-monitoring-issues.adoc
+
+:_mod-docs-content-type: PROCEDURE
+[id="resolving-the-kubepersistentvolumefillingup-alarm-firing-for-prometheus_{context}"]
+= Resolving the KubePersistentVolumeFillingUp alarm firing for Prometheus
+
+As a cluster administrator, you can resolve the `KubePersistentVolumeFillingUp` alarm being triggered for Prometheus. 
+
+The alert fires when a persistent volume (PV) claimed by a `prometheus-k8s-*` pod in the `openshift-monitoring` project has less than 3% of its total space remaining. This can cause Prometheus to not function normally.
+
+You can resolve this issue by freeing some space for the PV by removing the three oldest Prometheus time series database (TSDB) blocks.
+
+.Prerequisites
+
+ifndef::openshift-dedicated,openshift-rosa[]
+* You have access to the cluster as a user with the `cluster-admin` cluster role.
+endif::openshift-dedicated,openshift-rosa[]
+ifdef::openshift-dedicated,openshift-rosa[]
+* You have access to the cluster as a user with the `dedicated-admin` role.
+endif::openshift-dedicated,openshift-rosa[]
+* You have installed the OpenShift CLI (`oc`).
+
+.Procedure
+
+. Remove the three oldest Prometheus TSDB blocks to make more space available for the PV:
++
+[source,terminal]
+----
+$ oc debug <prometheus_k8s_pod_name> -n openshift-monitoring \// <1>
+-c prometheus --image=$(oc get po -n openshift-monitoring <prometheus_k8s_pod_name> \// <1>
+-o json | jq '.spec.containers[] |select(.name=="prometheus").image' -r) \
+-- sh -c 'ls -latr /prometheus/ | egrep -o "[0-9|A-Z]{26}" | head -3 | \
+while read BLOCK; do rm -rf /prometheus/$BLOCK; done'
+----
+<1> Replace `<prometheus_k8s_pod_name>` with the pod mentioned in the `KubePersistentVolumeFillingUp` alarm description.
+
+. Verify the usage of the mounted PV:
++
+[source,terminal]
+----
+$ oc debug <prometheus_k8s_pod_name> -n openshift-monitoring \// <1>
+--image=$(oc get po -n openshift-monitoring <prometheus_k8s_pod_name> \// <1>
+-o json | jq '.spec.containers[] |select(.name=="prometheus").image' -r) -- df -h /prometheus/
+----
+<1> Replace `<prometheus_k8s_pod_name>` with the pod mentioned in the `KubePersistentVolumeFillingUp` alarm description.
++
+.Example command for the `prometheus-k8s-0` pod
++
+[source,terminal]
+----
+$ oc debug prometheus-k8s-0 -n openshift-monitoring \
+--image=$(oc get po -n openshift-monitoring prometheus-k8s-0 \
+-o json | jq '.spec.containers[] |select(.name=="prometheus").image' -r) -- df -h /prometheus/
+----
+.Example output
++
+[source,terminal]
+----
+Starting pod/prometheus-k8s-0-debug-j82w4 ...
+Filesystem      Size  Used Avail Use% Mounted on
+/dev/nvme0n1p4  40G   15G  40G  37% /prometheus
+
```

Comment:
I need some guidance for this output on what exactly am I looking for. 
Is this the usage of the mounted PV? So the usage would have to be lower than 97% for the alarm to stop firing? Do I understand this correctly?

Reply:
please note there are 2 KubePersistentVolumeFillingUp alert, one is critical alert which the available PV is less than 3% of the total PV size(in other words, PV usage is >97%), one is warning alert and the current  available PV is less than 15% and is expected to fill up within four days
```
$ oc -n openshift-monitoring get prometheusrules kubernetes-monitoring-rules -o yaml
...
    - alert: KubePersistentVolumeFillingUp
      annotations:
        description: The PersistentVolume claimed by {{ $labels.persistentvolumeclaim
          }} in Namespace {{ $labels.namespace }} is only {{ $value | humanizePercentage
          }} free.
        runbook_url: https://github.com/openshift/runbooks/blob/master/alerts/cluster-monitoring-operator/KubePersistentVolumeFillingUp.md
        summary: PersistentVolume is filling up.
      expr: |
        (
          kubelet_volume_stats_available_bytes{namespace=~"(openshift-.*|kube-.*|default)",job="kubelet", metrics_path="/metrics"}
            /
          kubelet_volume_stats_capacity_bytes{namespace=~"(openshift-.*|kube-.*|default)",job="kubelet", metrics_path="/metrics"}
        ) < 0.03
        and
        kubelet_volume_stats_used_bytes{namespace=~"(openshift-.*|kube-.*|default)",job="kubelet", metrics_path="/metrics"} > 0
        unless on(namespace, persistentvolumeclaim)
        kube_persistentvolumeclaim_access_mode{namespace=~"(openshift-.*|kube-.*|default)", access_mode="ReadOnlyMany"} == 1
        unless on(namespace, persistentvolumeclaim)
        kube_persistentvolumeclaim_labels{namespace=~"(openshift-.*|kube-.*|default)",label_alerts_k8s_io_kube_persistent_volume_filling_up="disabled"} == 1
      for: 1m
      labels:
        severity: critical
    - alert: KubePersistentVolumeFillingUp
      annotations:
        description: Based on recent sampling, the PersistentVolume claimed by {{
          $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} is
          expected to fill up within four days. Currently {{ $value | humanizePercentage
          }} is available.
        runbook_url: https://github.com/openshift/runbooks/blob/master/alerts/cluster-monitoring-operator/KubePersistentVolumeFillingUp.md
        summary: PersistentVolume is filling up.
      expr: |
        (
          kubelet_volume_stats_available_bytes{namespace=~"(openshift-.*|kube-.*|default)",job="kubelet", metrics_path="/metrics"}
            /
          kubelet_volume_stats_capacity_bytes{namespace=~"(openshift-.*|kube-.*|default)",job="kubelet", metrics_path="/metrics"}
        ) < 0.15
        and
        kubelet_volume_stats_used_bytes{namespace=~"(openshift-.*|kube-.*|default)",job="kubelet", metrics_path="/metrics"} > 0
        and
        predict_linear(kubelet_volume_stats_available_bytes{namespace=~"(openshift-.*|kube-.*|default)",job="kubelet", metrics_path="/metrics"}[6h], 4 * 24 * 3600) < 0
        unless on(namespace, persistentvolumeclaim)
        kube_persistentvolumeclaim_access_mode{namespace=~"(openshift-.*|kube-.*|default)", access_mode="ReadOnlyMany"} == 1
        unless on(namespace, persistentvolumeclaim)
        kube_persistentvolumeclaim_labels{namespace=~"(openshift-.*|kube-.*|default)",label_alerts_k8s_io_kube_persistent_volume_filling_up="disabled"} == 1
      for: 1h
      labels:
        severity: warning
```

Reply:
> I need some guidance for this output on what exactly am I looking for. Is this the usage of the mounted PV? So the usage would have to be lower than 97% for the alarm to stop firing? Do I understand this correctly?

yes, for the critical KubePersistentVolumeFillingUp alert, if the mounted PV usage is <97%, the alert will stop firing

Reply:
Hello @juzhao, thank you for the information and the command! (I will save it so that I can use it in the future). This answers pretty much all my questions, thank you!